### PR TITLE
fix(cloud): validate resource scope against cloud token (GKO-2961)

### PR DESCRIPTION
**Pull request by benoitgravitee**
## Issue

https://gravitee.atlassian.net/browse/GKO-2961

## Summary

- Add an HTTP transport that validates `organization_id` and `environment_id` from outgoing API URLs against Gravitee Cloud token claims.
- Extend cloud initialization to expose parsed token claims and share scope validation with provider configure.
- Block apply-time requests when a resource targets an environment or organization not permitted by the cloud token, without modifying generated resource CRUD code.

## Test plan

- [x] `go test ./internal/provider/...`
- [x] Confirm `terraform apply` fails when `cloud_auth` is scoped to env A and a resource sets `environment_id` to env B
- [x] Confirm self-hosted (`bearer_auth`) behavior is unchanged

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -201,7 +202,10 @@ func (p *ApimProvider) Configure(ctx context.Context, req provider.ConfigureRequ
 	}
 
 	httpClient := http.DefaultClient
-	httpClient.Transport = NewProviderHTTPTransport(providerHTTPTransportOpts)
+	httpClient.Transport = NewCloudScopeTransport(
+		NewProviderHTTPTransport(providerHTTPTransportOpts),
+		cloudInit.Claims,
+	)
```
[the remark this answers is not in the document]
Comment (benoitgravitee) [reply internal/provider/provider.go:211]: It is actually the intent

File: internal/provider/gravitee_cloud.go [hunk 2]
```diff
@@ -34,33 +40,46 @@ func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProvider
...
+	} else if err := validateCloudScope(jwtData, configuredOrgID, data.EnvironmentID.ValueString()); err != nil {
+		resp.Diagnostics.AddError("Cloud Token misconfiguration", err.Error())
+		return CloudInitResult{ServerURL: serverUrl}
+	}
+
+	claims := jwtData
+	result := CloudInitResult{
+		ServerURL: serverUrl,
+		Claims:    &claims,
 	}
```
[the remark this answers is not in the document]
Comment (benoitgravitee) [reply internal/provider/gravitee_cloud.go:63]: I prefer a copy, that is safer

File: internal/provider/cloud_scope_transport.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	return &cloudScopeTransport{
+		transport: transport,
+		claims:    claims,
+	}
+}
+
+func (t *cloudScopeTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	orgID, envID, ok := scopeFromRequestPath(req.URL.Path)
+	if ok {
+		if err := validateCloudScope(*t.claims, orgID, envID); err != nil {
```
Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: The transport reuses `validateCloudScope`, which skips validation whenever the value is the literal string `DEFAULT`. At request time `DEFAULT` is not a sentinel for "unset" — it's a real path segment, and it's the org/env id every self-hosted config uses, so it's exactly what users carry over when moving a config to `cloud_auth`.

Verified empirically: with a valid cloud token, `PUT /apim/automation/organizations/DEFAULT/environments/DEFAULT/groups/x` and `.../organizations/<token-org>/environments/DEFAULT/groups/x` both pass validation and are sent to CloudGate, while `.../organizations/wrong/...` is correctly rejected. A resource with `organization_id = "DEFAULT"` or `environment_id = "DEFAULT"` therefore escapes the scope check this PR introduces.

File: internal/provider/gravitee_cloud.go [hunk 4]
```diff
@@ -15,16 +15,22 @@ const cloudGateUrlTemplate = "https://%s.cloudgate.gravitee.io/apim/automation"
...
-// CloudInitializer setup url/env/org if required. Keeps non default value intact
-func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProviderModel, resp *tfp.ConfigureResponse) string {
+// CloudInitResult holds provider configuration derived from a Gravitee Cloud token.
+type CloudInitResult struct {
+	ServerURL string
+	Claims    *CloudTokenClaimsData
+}
+
+// CloudInitializer setup url/env/org if required. Keeps non default value intact.
+func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProviderModel, resp *tfp.ConfigureResponse) CloudInitResult {
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: `CloudInitializer`: no test for an explicitly configured `environment_id` that IS present in the token's `envs` (the accept-side of the line-46 else-leg). Every existing test uses either `DEFAULT` or an invalid value, so the accept path of that branch never runs — and it's the only remedy the "environment_id is required in that case" error offers for a multi-environment token.

Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: `CloudInitializer`: no test for an explicitly configured `organization_id` that matches the token's `org` claim (the accept-side of the line-54 else-leg). Only `DEFAULT` and a mismatching org are covered, so a regression that rejects the correct org would go unnoticed.

File: internal/provider/cloud_scope_transport.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	return &cloudScopeTransport{
+		transport: transport,
+		claims:    claims,
+	}
+}
+
+func (t *cloudScopeTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	orgID, envID, ok := scopeFromRequestPath(req.URL.Path)
+	if ok {
```
Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: When `scopeFromRequestPath` returns `ok == false`, the scope check is silently bypassed and the request is forwarded unvalidated. Any request path that doesn't match `cloudScopePathPattern` (path shape change, escaped segment, a future org-scoped endpoint) leaves no trace that enforcement was skipped, so a live enforcement gap would be invisible in logs.

Verified against the current SDK: every generated operation path does include `/organizations/{id}/environments/{id}`, so this doesn't fire today — but this transport wraps the process-global `http.DefaultClient` and intercepts every outgoing request, so any future endpoint without org/env in its path would silently bypass enforcement with zero observability.

File: internal/provider/gravitee_cloud.go [hunk 6]
```diff
@@ -34,33 +40,46 @@ func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProvider
...
 	}
 
 	// returned computed URL
-	return jwtData.baseUrl()
+	result.ServerURL = jwtData.baseUrl()
+	return result
+}
+
+func validateCloudScope(claims CloudTokenClaimsData, orgID, envID string) error {
+	if envID != "" && envID != "DEFAULT" && !slices.Contains(claims.Envs, envID) {
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: `validateCloudScope` is shared between config-time (`CloudInitializer`) and request-time (`cloudScopeTransport.RoundTrip`) but exempts the literal value `"DEFAULT"` unconditionally. At config time that's correct — `DEFAULT` is the provider's sentinel for "unset". At request time the values are concrete URL path segments: a resource with `organization_id = "DEFAULT"` or `environment_id = "DEFAULT"` yields `/organizations/DEFAULT/environments/.../` which the transport waves through unvalidated.

File: internal/provider/gravitee_cloud_test.go [hunk 7]
```diff
@@ -66,8 +68,8 @@ func TestCloudInitializerValidTokenAndNonDefaultURL(t *testing.T) {
 	}
 
 	serverUrl := "http://localhost"
-	url := CloudInitializer(auth, serverUrl, data, resp)
-	assert.Equal(t, url, serverUrl)
+	result := CloudInitializer(auth, serverUrl, data, resp)
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud_test.go]: `CloudInitializer`'s `Claims` field is what decides whether `provider.go` installs the scope-enforcing transport, and its content (org/envs) is what every request is later validated against. `TestCloudInitializerValidTokenAndNonDefaultURL` is a success-path test with a valid cloud token, yet it never asserts anything about `result.Claims`.

Same gap in `TestCloudInitializerValidTokenNonDefaultGeography` (line 96).

File: internal/provider/gravitee_cloud_test.go [hunk 8]
```diff
@@ -91,8 +93,8 @@ func TestCloudInitializerValidTokenNonDefaultGeography(t *testing.T) {
 		Diagnostics: make(diag.Diagnostics, 0),
 	}
 
-	url := CloudInitializer(auth, defaultCloudUrl, data, resp)
-	assert.Equal(t, fmt.Sprintf(cloudGateUrlTemplate, "us"), url)
+	result := CloudInitializer(auth, defaultCloudUrl, data, resp)
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud_test.go]: Same gap as `TestCloudInitializerValidTokenAndNonDefaultURL` (line 71): this success-path test with a valid cloud token never asserts anything about `result.Claims`.

File: internal/provider/gravitee_cloud_test.go [hunk 9]
```diff
@@ -40,8 +41,9 @@ func TestCloudInitializerValidTokenAndConfig(t *testing.T) {
 		Diagnostics: make(diag.Diagnostics, 0),
 	}
 
-	url := CloudInitializer(auth, defaultCloudUrl, data, resp)
-	assert.Equal(t, url, defaultCloudUrl)
+	result := CloudInitializer(auth, defaultCloudUrl, data, resp)
+	assert.Equal(t, defaultCloudUrl, result.ServerURL)
+	assert.NotNil(t, result.Claims)
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud_test.go]: `assert.NotNil(t, result.Claims)` is a presence-only assertion on a value whose content IS the contract: the org, envs and cpg claims decoded from the token are exactly what the scope transport later compares every request against. The assertion passes even if `Claims` were populated with a zero-value `CloudTokenClaimsData{}`.

File: tests/acceptance/cloud_scope_test.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	cloudTestOrg  = "2806dece-d045-463d-86de-ced045963d84"
+	cloudTestEnv  = "9b5e326f-f68a-45f9-9e32-6ff68ae5f92c"
+	cloudTestEnv2 = "357c35e3-6635-443a-9812-312e2f548606"
+	cloudWrongOrg = "00000000-0000-0000-0000-000000000001"
+	cloudWrongEnv = "11111111-1111-1111-1111-111111111111"
+)
+
+// Verifies cloud token scope validation rejects misconfiguration before API calls.
+func TestCloudAuthScope_rejectsUnauthorizedScope(t *testing.T) {
+	t.Parallel()
```
Comment (carlos-andres-osorio) [inline tests/acceptance/cloud_scope_test.go]: This test calls `t.Parallel()` and runs in the same acceptance suite (run with `-p 2`) as every other test in `tests/acceptance`, all of which also call `t.Parallel()`. `Configure` mutates the process-global `http.DefaultClient`, so while this test is in flight it installs a scope-enforcing transport carrying its synthetic cloud claims into the client shared by the concurrently running `bearer_auth` tests against the live APIM — those tests would then get 'cloud token scope violation' failures for their own org/env. Contamination runs both ways: a sibling test's `Configure` can also overwrite this test's transport mid-flight.

File: tests/acceptance/testdata/TestCloudAuthScope_rejectsUnauthorizedScope/main.tf [hunk 11]
```diff
@@ -0,0 +1,42 @@
+provider "apim" {
+  cloud_auth      = var.cloud_token
```
Comment (carlos-andres-osorio) [inline tests/acceptance/testdata/TestCloudAuthScope_rejectsUnauthorizedScope/main.tf]: This batch's test config doesn't cover the case where self-hosted authentication (`bearer_auth`) is used — requests are still sent and `terraform apply` succeeds exactly as before this change. The provider block hardcodes `cloud_auth = var.cloud_token`, so no step using this testdata can exercise the non-cloud path, and there's no `bearer_auth` variant in the batch. (The nil-claims passthrough is unit-tested elsewhere — `TestCloudScopeTransportPassthroughWithoutClaims` — so this is a gap in this batch's acceptance-tier config coverage, not a total absence of coverage.)

File: internal/provider/gravitee_cloud.go (hunk 6 above)
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: Same location as the `DEFAULT` exemption above, different angle: the new `envID != ""` and `orgID != ""` guards also silently drop config-time validation that existed before this change. Previously an empty environment/organization id would produce a `Cloud Token misconfiguration` error; now it's accepted and the provider proceeds with an empty environment/organization.

Reachable without HCL gymnastics: `provider.go` uses `os.LookupEnv("APIM_ENV_ID")`, which returns `ok == true` for an exported-but-empty variable, so `APIM_ENV_ID=` (exported empty) sets `EnvironmentID` to `""` and bypasses both the `DEFAULT` branch and the scope check.

File: internal/provider/gravitee_cloud.go [hunk 12]
```diff
@@ -34,33 +40,46 @@ func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProvider
...
+		return result
 	}
 
 	// returned computed URL
-	return jwtData.baseUrl()
+	result.ServerURL = jwtData.baseUrl()
+	return result
+}
+
+func validateCloudScope(claims CloudTokenClaimsData, orgID, envID string) error {
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: Nit: `validateCloudScope(claims, orgID, envID string) error` takes two positional, identically-typed `string` params that are both opaque UUID-shaped identifiers. Nothing in the type system stops a future call site from passing them in the wrong order — this is the single security gate this PR introduces, reused at three call sites. Not required for this PR, but worth hardening if it's simple — e.g. distinct types (`type cloudOrgID string` / `type cloudEnvID string`) or a small named struct.

File: internal/provider/cloud_scope_transport.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package provider
```
Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: Nit: This adds a hand-written `http.RoundTripper` in `internal/provider` that intercepts every outgoing request and regex-parses the URL shape. Convention TPAA2 keeps HTTP/hooks in `internal/sdk`, which already exposes a hand-owned hook registration point for this. Not blocking, but worth considering for follow-up: `internal/sdk/internal/hooks/registration.go` as a before-request hook.

File: internal/provider/cloud_scope_transport.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+		transport: transport,
+		claims:    claims,
+	}
+}
+
+func (t *cloudScopeTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	orgID, envID, ok := scopeFromRequestPath(req.URL.Path)
+	if ok {
+		if err := validateCloudScope(*t.claims, orgID, envID); err != nil {
+			return nil, fmt.Errorf("cloud token scope violation: %w", err)
```
Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: Nit: `RoundTrip` returns an error without closing `req.Body`, which technically violates the `http.RoundTripper` contract ("RoundTrip must always close the body, including on errors"). No practical consequence today since the SDK always supplies in-memory buffers.

Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: Nit: `RoundTrip` rejects a request on scope violation with no `tflog` call, unlike the sibling transport it wraps (`providerHttpTransport`, which logs every outgoing request/response). A blocked request won't appear in the standard `TF_LOG=DEBUG` HTTP trace, making it harder to diagnose via that channel — though the error itself does surface to the end user via the Terraform diagnostic.

File: internal/provider/gravitee_cloud.go [hunk 15]
```diff
@@ -34,33 +40,46 @@ func CloudInitializer(auth shared.Security, serverUrl string, data *ApimProvider
...
-			return serverUrl
+			return CloudInitResult{ServerURL: serverUrl}
 		}
 		data.EnvironmentID = basetypes.NewStringValue(jwtData.Envs[0])
-	} else if !slices.Contains(jwtData.Envs, configuredEnvID) {
-		// if set, check env in token
-		resp.Diagnostics.AddError("Cloud Token misconfiguration",
-			fmt.Sprintf("cloud token does not contain environment [%s], it must be one of: %s", configuredEnvID, jwtData.Envs))
-		return serverUrl
+	} else if err := validateCloudScope(jwtData, data.OrganizationID.ValueString(), configuredEnvID); err != nil {
```
Comment (carlos-andres-osorio) [inline internal/provider/gravitee_cloud.go]: Nit: `validateCloudScope` failures inside `CloudInitializer` are converted to diagnostics with no ERROR-level trace of the fail-closed decision beyond the user-facing diagnostic — same observability gap as the transport-level rejection.

File: internal/provider/cloud_scope_transport.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
+package provider
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+)
+
+var cloudScopePathPattern = regexp.MustCompile(`/organizations/([^/]+)/environments/([^/]+)`)
```
Comment (carlos-andres-osorio) [inline internal/provider/cloud_scope_transport.go]: Nit: `cloudScopePathPattern` is unanchored and matches the first `/organizations/X/environments/Y` occurrence anywhere in the path. Works fine today since every current SDK operation path has this prefix right after the base path, but a future nested/proxied route could match mid-path and validate the wrong pair.

**Comment by carlos-andres-osorio** [comment]
This PR is missing the `docs-needed`/`no-docs` label. Can you confirm which applies and add it?

File: internal/provider/cloud_scope_transport.go (hunk 3 above)
Comment (benoitgravitee) [reply internal/provider/cloud_scope_transport.go]: I Cloud Token cannot have "DEFAULT" as an env. Not worth fixing this.

File: tests/acceptance/testdata/TestCloudAuthScope_rejectsUnauthorizedScope/main.tf (hunk 11 above)
Comment (benoitgravitee) [reply tests/acceptance/testdata/TestCloudAuthScope_rejectsUnauthorizedScope/main.tf]: Yes it is know issue. Bearer Auth is not recommended for end user to use with a Cloud Token, they would need to know the server URLs Org and Env, the purpose of Cloud Initialiser. For regular bearer auth, with a token, it is not tested here because it requires a complex setup. This is typically tested in e2e tests in another repo.

File: internal/provider/gravitee_cloud.go (hunk 6 above)
Comment (benoitgravitee) [reply internal/provider/gravitee_cloud.go]: This is an edge case unrelated to cloud initializer. It is a Speakeasy issue. provider.go should check for "" and set "DEFAULT" in that case. I don't want to add that to the generated code if I can avoid. I will report the issue though.

File: internal/provider/cloud_scope_transport.go (hunk 13 above)
Comment (benoitgravitee) [reply internal/provider/cloud_scope_transport.go]: Already validated as a no go with Speakeasy.

File: internal/provider/cloud_scope_transport.go (hunk 16 above)
Comment (benoitgravitee) [reply internal/provider/cloud_scope_transport.go]: There is no reason why this would happen, org and env are base entities, this unnecessary protection. Let's use KISS principle for this.

**Comment by benoitgravitee** [comment]
> This PR is still missing the `docs-needed`/`no-docs` label (flagged earlier on Aug 24, still unresolved). Can you confirm which applies and add it?

Added. Thanks for the review.

**Review by a-cordier — APPROVED** [review]
